cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/TaskWorkerHttpHandlerInternal.java

```diff
@@ -166,34 +188,94 @@ public TaskWorkerHttpHandlerInternal(CConfiguration cConf,
   private void enablePeriodicRestart(CConfiguration cConf,
       Consumer<String> stopper) {
     int duration = cConf.getInt(
-        Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION_SECOND, 0);
+        TaskWorker.CONTAINER_KILL_AFTER_DURATION_SECOND, 0);
     int lowerBound = (int) (duration - duration * DURATION_FRACTION);
     int upperBound = (int) (duration + duration * DURATION_FRACTION);
-    if (lowerBound > 0) {
-      int waitTime =
-          (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
-      Executors.newSingleThreadScheduledExecutor(
-              Threads.createDaemonThreadFactory("task-worker-restart"))
-          .scheduleWithFixedDelay(
-              () -> {
-                if (mustRestart.get()) {
-                  // We force pod restart as the ongoing request has not finished since last
-                  // periodic restart check.
-                  stopper.accept("");
-                  return;
-                }
-                // we restart once ongoing request (which has set runningRequestCount to 1)
-                // finishes.
-                mustRestart.set(true);
-                if (runningRequestCount.compareAndSet(0, 1)) {
-                  // there is no ongoing request. pod gets restarted.
-                  stopper.accept("");
-                }
-              },
-              waitTime,
-              waitTime,
-              TimeUnit.SECONDS);
+    if (lowerBound <= 0) {
+      return;
+    }
+    int waitTime = (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
+    Executors.newSingleThreadScheduledExecutor(
+            Threads.createDaemonThreadFactory("task-worker-restart"))
+        .scheduleWithFixedDelay(() -> {
+          if (mustRestart.get()) {
+            // We force pod restart as the ongoing request has not finished since last
+            // periodic restart check.
+            stopper.accept("");
+            return;
+          }
+          // we restart once ongoing request (which has set runningRequestCount to 1)
+          // finishes.
+          mustRestart.set(true);
```

Comment:
Why do we need both mustRestart and lameDock logic?

Reply:
Here, if there is an ongoing request, we can move the pod into lameDock.
Basically, I'm thinking of changing lameDock semantic from a task has timeout to pod needs to restart in the next safest time. WDYT?

Reply:
Changed to carry out a periodic restart using the existing periodic restart code. I had used the lameduck concept to ensure the service is restarted only when a task is stuck. In the periodic restart approach, we may be restarting task workers even when no task is stuck.